sonar-plugin-api/src/main/java/org/sonar/api/server/authentication/BaseIdentityProvider.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.api.server.authentication;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * @since 5.4
+ */
+public interface BaseIdentityProvider extends IdentityProvider {
+
+  /**
+   * Entry-point of authentication workflow. Executed by core when user
+   * clicks on the related button in login form (GET /sessions/init/{provider key}).
+   */
+  void init(Context context);
+
+  interface Context {
+
+    /**
+     * Get the received HTTP request.
+     * Note - {@code getRequest().getSession()} must not be used in order to support
+     * future clustering of web servers without stateful server sessions.
+     */
+    HttpServletRequest getRequest();
+
+    /**
+     * Get the HTTP response to send
+     */
+    HttpServletResponse getResponse();
+
+    /**
+     * Return the server base URL
+     */
+    String getServerBaseURL();
+
+    /**
+     * Authenticate and register the user into the platform
```

Comment:
Way to handle errors should be mentioned in Javadoc : what about technical errors (IllegalStateException ?) and functional errors (conflicting email ?). Moreover it should mention that user is registered during the first authentication then his information are updated (name and email)